### ISSUE TITLE
Public API surface for final classes should not have virtual functions, unless they are overriding virtual functions in a base class.

### ISSUE THREAD

**Issue by ahsonkhan** [Azure/azure-sdk-for-cpp#5253]
From https://apiview.dev/Assemblies/Conversation/7210656fdab948559a0a3aa0b3473953#92efa77b9fe24ca994a49c672ee4afda

For example, these methods, marked as virtual used in mocking for tests, under an #ifdef, still leak the fact that the method is (protected) virtual in the public surface area, observable to APIView and end users:
https://github.com/Azure/azure-sdk-for-cpp/blob/e69559f45e9e717eba3421ee8ac3d58a3fb13c87/sdk/identity/azure-identity/inc/azure/identity/azure_cli_credential.hpp#L109-L115

We can consider moving the virtual keyword into the testing ifdef.

From @LarryOsterman:
> We 100% are going to run into issues with linters who will correctly diagnose virtual functions and/or protected fields in final classes as errors.

cc @antkmsft

**Comment by antkmsft** [comment]
I think it is risky to have ifdefs on virtual.
If you don't like other options, then let's make the class to have std::functions as member, and then make the code to call them ifdef testing. But do not conditionally compile the std::function members either - that is also risky. Only conditionally compile their invocation.
And then to be able to set these std::functions from test code, we will need to add test class as friend.

**Comment by antkmsft** [comment]
We will eventually get the same thing, maybe even less performant, with more lines of code to support, and with more potential points of failure. I think conditional final + virtual functions is lesser of evils.

**Comment by antkmsft** [comment]
One more comment "correctly diagnose virtual functions and/or protected fields in final classes as errors" - these are not errors! Language perfectly allows it. It is a problem of the linters then. I would disable these lines with pragmas then. But so far, this approach is being used in 5 places, since 2021, and we haven't heard about linters failing.

**Comment by antkmsft** [comment]
From conversation with Larry today: "https://www.godbolt.org/z/dPoYoxW7K. MSVC and gcc use `final` as a hint to remove the `virtual` keyword. Classes with final directly call the method. The method still shows up in the vtable but it's not used."

I.e. it removes the concern with the virtual call cost.

We chatted more, I provided some examples how conditional virtual can blow up pretty badly - yes, maybe when the build flags are not aligned, but it is possible to mess them up, and the resulting error is pretty fatal and hard to debug.

Larry said, "I still think that virtual in a final is sloppy and ugly but given that the compilers appear to ignore virtual when dealing with a final class, I'll concede that one."

Larry has also said, "I'm still going to push on protected though."

That we can do. So, we should not be making these virtuals preprocessor conditionals, but we should make them private. That can be done. I don't have issues with that.

Larry has also said that we should avoid/get rid of the practice of protected data members, which I also agree with, but that is a separate issue.

In general, I think it is also possible to think of more general approach for mocking and exposing the testing methods/data from the classes, and that is also a separate item: #5147.

**Comment by antkmsft** [comment]
TLDR: To close this GitHub issue, declare these virtual test hooks in "ifndef(TESTING_BUILD) final" classes as `private`.

**Comment by LarryOsterman** [comment]
One other thought: We already have precedents for declaring test hooks as friend classes, which should render the `private`/`protected` discussion moot.

**Comment by antkmsft** [comment]
@LarryOsterman, so if I understand correctly, we just close this item then?

**Comment by ahsonkhan** [comment]
Having friend classes for testing private implementation details (as a testing pattern) and having virtual methods to mock private implementation detail are separate topics.

**Comment by ahsonkhan** [comment]
Re-opening until existing API surface is fixed.
Follow up: https://github.com/Azure/azure-sdk-for-cpp/issues/5334
